flow/WriteOnlySet.actor.cpp

```diff
@@ -0,0 +1,232 @@
+
+// testing code
+namespace {
+
+// Some statistics
+std::atomic<unsigned long> instanceCounter = 0;
+std::atomic<unsigned long> numInserts = 0;
+std::atomic<unsigned long> numErase = 0;
+std::atomic<unsigned long> numLockedErase = 0;
+std::atomic<unsigned long> numCopied = 0;
+
+// A simple object that counts the number of its instances. This is used to detect memory leaks.
+struct TestObject {
+	mutable std::atomic<unsigned> _refCount = 1;
+	TestObject() { instanceCounter.fetch_add(1); }
+	void delref() const {
+		if (--_refCount == 0) {
+			delete this;
+			--instanceCounter;
+		}
+	}
+	void addref() const { ++_refCount; }
+};
+
+using TestSet = WriteOnlySet<TestObject, unsigned, 128>;
+using Clock = std::chrono::steady_clock;
+
+// An actor that can join a set of threads in an async way.
+ACTOR Future<Void> threadjoiner(std::shared_ptr<std::vector<std::thread>> threads, std::shared_ptr<TestSet> set) {
+	loop {
+		wait(delay(0.1));
+		for (unsigned i = 0;;) {
+			if (threads->size() == i) {
+				break;
+			}
+			auto& t = (*threads)[i];
+			if (t.joinable()) {
+				t.join();
+				if (i + 1 < threads->size()) {
+					std::swap(*threads->rbegin(), (*threads)[i]);
+				}
+				threads->pop_back();
+			} else {
+				++i;
+			}
+		}
+		if (threads->empty()) {
+			set->copy();
+			ASSERT(instanceCounter.load() == 0);
+			return Void();
+		}
+	}
+}
+
+// occasionally copy the contents of the past set.
+void testCopier(std::shared_ptr<TestSet> set, std::chrono::seconds runFor) {
+	auto start = Clock::now();
+	while (true) {
+		if (Clock::now() - start > runFor) {
+			return;
+		}
+		auto copy = set->copy();
+		numCopied.fetch_add(copy.size());
+		std::this_thread::sleep_for(std::chrono::milliseconds(10));
+	}
+}
+
+// In a loop adds and removes a set of objects to the set
+void writer(std::shared_ptr<TestSet> set, std::chrono::seconds runFor) {
+	auto start = Clock::now();
+	std::random_device rDev;
+	DeterministicRandom rnd(rDev());
+	while (true) {
+		unsigned inserts = 0, erases = 0;
+		if (Clock::now() - start > runFor) {
+			return;
+		}
+		std::vector<TestSet::Index> positions;
+		for (int i = 0; i < rnd.randomInt(1, 101); ++i) {
+			Reference<TestObject> o(new TestObject());
+			auto pos = set->insert(o);
+			if (pos == TestSet::npos) {
+				// could not insert -- ignore
+				break;
+			}
+			++inserts;
+			ASSERT(pos < TestSet::capacity);
+			positions.push_back(pos);
+		}
+		rnd.randomShuffle(positions);
+		for (auto p : positions) {
+			if (!set->erase(p)) {
+				++numLockedErase;
+			}
+			++erases;
+		}
+		numInserts.fetch_add(inserts);
+		numErase.fetch_add(erases);
+		ASSERT(inserts == erases);
+		std::this_thread::sleep_for(std::chrono::milliseconds(1));
+	}
+}
+
+// This unit test creates 5 writer threads and one copier thread.
+TEST_CASE("/flow/WriteOnlySet") {
+	if (g_network->isSimulated()) {
+		// This test is not deterministic, so we shouldn't run it in simulation
+		return Void();
+	}
+	auto set = std::make_shared<TestSet>();
```

Comment:
I don't think we technically need shared ownership here since the set can just outlive all the threads, but this is fine.

Reply:
This is me being lazy... The shared-pointer makes the code slightly simpler to read/write (and reasoning about memory lifetime). I think for test-code this is ok.